Dockerize

ichicho opened the pull request:
Your project is very nice and I managed to dockerize it(#23 ).

ichicho committed:
create basic Dockerfile with chrome, webdriver and selenium installed

ichicho committed:
change variable case

ichicho committed:
add comments

ichicho committed:
Complete code part of dockerization

Add -d option for docker usage in nanaco_auto_fill.py and
change browser to chromium.

ichicho committed:
Dockerize readme (#4)

* Update README.md

* Update README.md

* Update README.md

* Update README.md

* Update README.md

ichicho commented:
> Your project is very nice and I managed to dockerize your project.

ichicho commented:
Reopen because I misclicked to close it.
PS:
I think my docker image can be run on Windows. However, I have never used docker on Windows. I guess some commands should be adjusted for PowerShell.

kadoyau commented:
Awesome PR!
I'm sorry I haven't noticed. I'll check this PR ASAP.

kadoyau commented:
[In progress] I'm trying to use it on Windows.

ichicho commented:
Please contact me if you need any help.

kadoyau commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+# Python3 Debian image
+FROM python:3
```
ref. https://hub.docker.com/_/python

kadoyau commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+    apt install -y --no-install-recommends \
+                curl \
+                unzip \
+                chromium && \
```
Is python:3 an official Debian/Ubuntu image?
if it's not, `apt-get clean` keeps the image size smaller. doesn't it?
>Official Debian and Ubuntu images automatically run apt-get clean, so explicit invocation is not required.
https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run
 
```suggestion
                chromium && \
                apt-get clean && \
```

ichicho replied:
The base image is supposed to be an official Debian release(I used python:3.9.0-buster for testing). I will add a Debian tag for the base Python3 image. Therefore,  I guess `apt-get clean` isn't necessary and `rm -rf /var/lib/apt/lists/*` is enough. If I'm wrong, please feel free to correct me.

ichicho replied:
The good news is we confirmed nagias can be run in docker. However, the image size is 1.3GB now. I will try to switch base image to alpine to make the whole image lighter.

kadoyau replied:
Sorry, I don't know the underlying meaning of what documents say.
I have dug the docker code but I couldn't figure out how apt-clean is called. (Notice: I'm a Docker beginner)
I simply thought the documents literally referring to Ubuntu/Debian package only.

By the way, Alpine is better than Debian:+1:

Memo:
`python:3` use Debian package
https://hub.docker.com/_/python
https://hub.docker.com/layers/python/library/python/3/images/sha256-db968378628299a9164f0ac38bf46d257ace7365e31b2ded3a7e9f3eed4ba497?context=explore
https://github.com/docker-library/python/

ichicho replied:
For Ubuntu/Debian, just run `rm -rf /var/lib/apt/lists/*` after every `apt-get install`(in same `RUN` command). `apt-get clean` will be automatically conducted in official image.
For other Linux distribution image, to clear cache, command should be adjusted.

>Memo:
>`python:3` use Debian package

This is my bad. `python:3` is ambiguous. Using a detailed tag is better.

kadoyau commented on nanaco_auto_fill.py:
```diff
@@ -32,10 +33,15 @@ def __init__(self, login_type, use_canary, is_quiet):
+        if self.__is_docker:
+            options.binary_location = '/usr/bin/chromium'
+            options.add_argument('--headless')
+            options.add_argument('--no-sandbox')
```
Although we should use the `--security-opt` instead of the `--no-sandbox`, It's OK as I think the site is safe.
ref.
https://github.com/armbues/chrome-headless/blob/master/README.md#headless-google-chrome
https://hub.docker.com/r/zenika/alpine-chrome/
- >The best: With seccomp

ichicho replied:
Sorry that I know little about web engineering. If it is better, let's just use it. Could you please help me with this issue?

Can we use the security profile suggested in the link you show me? or some changes are required?
Link: https://github.com/armbues/chrome-headless/blob/master/README.md#headless-google-chrome
>Download @jessfraz Docker security profile for Chrome:
>`wget https://raw.githubusercontent.com/jessfraz/dotfiles/master/etc/docker/seccomp/chrome.json`

kadoyau replied:
It sounds good:+1:

kadoyau reviewed (approved):
[![Image from Gyazo](https://i.gyazo.com/cc06f8b1c48dee1776c56a78ed7d2801.png)](https://gyazo.com/cc06f8b1c48dee1776c56a78ed7d2801)
This PR is working on Windows:heart:

kadoyau commented:
close #23

kadoyau commented:
Fixes #27

ichicho committed:
Switch base image to Alpine & no longer use --no-sandbox option (#5)

* Switch to Alpine

* Use seccomp for better security

* Update README.md

ichicho commented:
I switched the base image from Debian to Alpine and the whole image is 443MB now :tada:. Please have a try.

Besides, to avoid `--no-sandbox`, a non-root user is added to run Chromium.
As for `--security-opt`, I copied the [chrome.json](https://github.com/jessfraz/dotfiles/blob/master/etc/docker/seccomp/chrome.json) to the repository and added `--security-opt seccomp=$PWD/chrome.json` to `docker run`. Even though I still have not fully understood Seccomp, it seems to work without any problems now :sweat_smile:.I would appreciate if you can check whether Seccomp profile is working properly.

kadoyau commented on README.md:
```diff
@@ -65,7 +65,76 @@ python nanaco_auto_fill.py -h
+
+## Dockerのセキュリティ
+- [Seccomp security profilesのDocker公式解説](https://docs.docker.com/engine/security/seccomp/)
+- [Chrome Headlessを安全に使うために](https://github.com/Zenika/alpine-chrome#-the-best-with-seccomp)
```
👍

Commit:
Download a Chrome.json when user run it

As the profile might change, download it when user run commands

ichicho committed:
Update info about testing environment

kadoyau commented:
I made a few changes and confirmed to work on windows.
I'll merge this PR if you're OK with it.

ichicho commented:
Thank you very much. Let's go! :rocket:

kadoyau commented:
Sorry to keep you waiting for so long!
Thank you for the PR!

ichicho commented:
Never mind. This is a good timing because it's about time to pay taxes. :scream: Your awesome project is very helpful.
